### DISCUSSION TITLE
[PHP-CS-Fixer] How to apply fixes (on risky rules)

### DISCUSSION THREAD

**Discussion by llaville** [oxsecurity/megalinter#3973] [Show and tell]
If you are a senior developer and know well concept of PHP-CS-Fixer, you probably know that fixing rules is simple. 

## Standard Rules

1. Checks that your source code follows your coding standard defined by your config file  `.php-cs-fixer.php` or whatever other name 

```shell
php-cs-fixer fix --dry-run --verbose --diff --config .php-cs-fixer.php
```

2. Once your are sure to apply fixes, you've just to remove the `--dry-run` option

```shell
php-cs-fixer fix --verbose --diff --config .php-cs-fixer.php
```

But it only works for standard rules (not risky rules).

## Risky Rules

If you want to follows (for example)  this `.php-cs-fixer.risky.php` config file : 

```php
<?php

$finder = (new PhpCsFixer\Finder())
    ->in('.')
;

return (new PhpCsFixer\Config())
    ->setRules([
        '@PER-CS' => true,
        '@PhpCsFixer:risky' => true,
    ])
    ->setFinder($finder)
;
```
 
And checks your source code with following command : 

```shell
php-cs-fixer fix --dry-run --verbose --diff --config .php-cs-fixer.risky.php
```

You'll get such Exception 

![risky_rules_exception](https://github.com/user-attachments/assets/3a5dc806-0d8e-40b6-960a-417b5042e89c)

To avoid such issue, you need to add extra arguments to your `.mega-linter.yml` config file, like that : 

```yaml
PHP_PHPCSFIXER_CONFIG_FILE: .php-cs-fixer.risky.php
PHP_PHPCSFIXER_ARGUMENTS:
  - "--allow-risky=yes"
  - "--diff"
```

And the results may be look like to :

![fixed](https://github.com/user-attachments/assets/210b4150-7c87-484e-8c92-b143f94d2cb3)

**Comment by echoix** [comment]
Great overview! Is it limited to diff in the console, or could be applied to the repo that is ran by megalinter too?

**Reply by llaville** [reply]
No it's not limited to console (and of course the `--diff` to show what has changed is optional)

**Comment by nvuillam** [comment]
@llaville amazing doc :)
Maybe you could add it, or a link to this discussion, in PHPCS MegaLinter doc ? ^^
